### ISSUE TITLE
exclamation points expand even inside quotation marks

### ISSUE THREAD

**Issue by Taywee** [nushell/nushell#13577]
### Describe the bug

If I try to run a command with an exclamation point followed by an integer inside a string, it will be automatically substituted.  I can't find a way to prevent this from happening, and I don't think it should happen inside strings anyway.

### How to reproduce

Try to execute `perl -e 'my $x = !0; print($x);'` in Nushell.  It automatically expands the `!0` in this situation.

### Expected behavior

Do not expand inside strings.

### Configuration

| key                | value                                                      |
| ------------------ | ---------------------------------------------------------- |
| version            | 0.96.1                                                     |
| major              | 0                                                          |
| minor              | 96                                                         |
| patch              | 1                                                          |
| branch             |                                                            |
| commit_hash        |                                                            |
| build_os           | linux-x86_64                                               |
| build_target       | x86_64-unknown-linux-gnu                                   |
| rust_version       | rustc 1.80.0 (051478957 2024-07-21) (Fedora 1.80.0-1.fc40) |
| cargo_version      | cargo 1.80.0 (376290515 2024-07-16)                        |
| build_time         | 2024-08-01 00:00:00 +00:00                                 |
| build_rust_channel | release                                                    |
| allocator          | mimalloc                                                   |
| features           | default, sqlite, system-clipboard, trash                   |
| installed_plugins  | bexpand 1.3.9, gstat 0.96.0                                |

**Comment by NotTheDr01ds** [comment]
I agree with this.  While I can think of use-cases for it inside quotes, those use-cases can be filled by using `history` inside a interpolated subexpression rather than a history substitution.

Note (not that this is helpful, but) the expansion only seems to take place when preceded by whitespace. `'!!'`, for instance, is not expanded, nor is `a!! `, etc.

**Comment by sholderbach** [comment]
Yes this bug is the result of the `!!` etc. expansion being directly handled in [reedline](https://github.com/nushell/reedline) (with the `bashism` feature) and it being not directly aware of the full syntax of Nushell (or any other language people may be using with it).

This should change to being something which we implement on top of a syntactical analysis inside nushell with some hook/mechanism for expansion in reedline.

- https://github.com/nushell/reedline/issues/556
- https://github.com/nushell/reedline/issues/311
